updater/lib/dependabot/updater/operations/update_all_versions.rb

```diff
@@ -48,10 +48,16 @@ def dependencies
             return []
           end
 
+          dependencies = if dependency_snapshot.ungrouped_dependencies.any?
+                           dependency_snapshot.ungrouped_dependencies
+                         else
+                           dependency_snapshot.allowed_dependencies
+                         end
```

Comment:
This is a pretty elegant solution - we could also push this branching up into the `dependency_snapshot so this class never needs to care?

i.e.

```
# dependency_snapshot.rb
def ungrouped_dependencies
  return allowed_dependencies unless dependency_groups.any?

  ...
end
```

Reply:
I used `ungrouped_dependencies.any?` as a [heuristic](https://github.com/dependabot/dependabot-core/pull/7110/files#diff-38103c3beda18d03beffc9179b8096d1023c737a04928271aed8d05ac7c18adcR70) on whether to run the ungrouped dependency updates 

So I don't think we can push this branching into `DependencySnapshot#ungrouped_dependencies` because the heuristic would return allowed_dependencies when the ungrouped dependency update should really be skipped

We could check `Experiments.enabled?(:grouped_updates_prototype)` as the heuristic instead, but once the feature flag is removed there won't be an easy way to check whether grouped updates are enabled based on a Snapshot.

Open to other options though!

Reply:
> So I don't think we can push this branching into DependencySnapshot#ungrouped_dependencies because the heuristic would return allowed_dependencies when the ungrouped dependency update should really be skipped

Ah I think I was getting confused with a previous version of the `applies_to?` method where we checked if `dependency_groups` were present as well as the experiment, so I was thinking we'd never get as far as that if we had no groups.

We can punt on this for now as I think the `DependencySnapshot` methods will need a semantics pass as I've started looking a little bit at a rebase strategy for single groups - it might more sense to revise this once we have a complete picture of create/rebase if it makes sense.